Iter15

GLZG opened the pull request.

Commit:
.

Commit:
.

Commit:
add dt

Commit:
fix

Commit:
.

Commit:
fix

Commit:
.

Commit:
.

Commit:
.

Commit:
.

GLZG committed:
middleware owerrite responcewriter

Commit:
.

Commit:
.

Commit:
.

Commit:
fix

Commit:
fix

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
fix

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
database dsn flag

Commit:
add ping handler

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
выбор хранилища в Init

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
Merge remote-tracking branch 'origin/iter15' into iter15

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

ekubyshin commented on internal/auth/auth.go:
```diff
@@ -0,0 +1,69 @@
+	}
+
+	if !token.Valid {
+		fmt.Println("token is not valid")
```
У тебя есть логгер в приложении, подобные желательно логировать через логгер, чтобы все было в одном стиле. Далее по коду есть подобные места.
Предлагаю исправить это.

ekubyshin commented on internal/auth/auth.go:
```diff
@@ -0,0 +1,69 @@
+	"github.com/golang-jwt/jwt/v4"
+)
+
+var userid int
```
Подскажи, какое назначение у этой переменной?

GLZG replied:
Я ввел ее для того, чтобы каждый раз, когда приходит запрос без токена, при формировании для него свежего JWT добавлять userid, после чего делать инкремент этой переменной
Таким образом, каждый запрос без токена получал бы в ответ jwt c индивидуальным userid
Это позволило бы сохранять шорты от несколькиз пользователей и вообще по сути являлось бы верным.

Я сделал это в коммите e112a7f и он не прошел тесты

Я сделал UserID  константой. Любой запрос без токена - неважно откуда и от кого он пришел - получит один и тот же userid. Это по сути не является правильным решением. И это прошло тесты: коммит 7d4763d

На данный момент я оставил ту реализацию, при которой тесты выполняются - это константный userid

Переменная userid осталась незадействованной. Могу ее убрать.

GLZG replied:
Ощущаю потребность проконсультироваться по этому моменту

ekubyshin commented on internal/auth/auth.go:
```diff
@@ -0,0 +1,69 @@
+	UserID int
+}
+
+const TOKENEXP = time.Hour * 3
```
Предлагаю придерживаться код стайла и объединять объявление констант и переменных. Это улучшает визуальное восприятие.
На мой взгляд, подобные константы желательно выносить в конфигурацию и брать оттуда, чтобы при необходимости можно было легко изменять непосреддственно при запуске приложения

GLZG replied:
Переместил TOKENEXP и SECRETKEY в config

ekubyshin commented on internal/auth/auth.go:
```diff
@@ -0,0 +1,69 @@
+const SECRETKEY = "supersecretkey"
+
+// BuildJWTString создаёт токен и возвращает его в виде строки.
+func BuildJWTString() (string, error) {
```
В курсе уже проходили продвинутую работу с ошибками, предалагаю, чтобы возвзвращались ошибки константы, чтобы потом можно было обрабатывать тип ошибки и писать определенную логику, там где это необходимо.

GLZG replied:
Я обьявил ошибку

// BuildJWTString указывает на ошибку формирования JWT
var ErrBuildJWTString error = errors.New("ошибка формирования JWT")

Если произойдет ошибка при формировании JWT, буду обьединять BuildJWTString с ней и возвращать

return "", errors.Join(ErrBuildJWTString, err)

ekubyshin commented on internal/auth/auth.go:
```diff
@@ -0,0 +1,69 @@
+	token, err := jwt.ParseWithClaims(tokenString, claims,
+		func(t *jwt.Token) (interface{}, error) {
+			if _, ok := t.Method.(*jwt.SigningMethodHMAC); !ok {
+				return nil, fmt.Errorf("unexpected signing method: %v", t.Header["alg"])
```
Аналогичное, предлагаю по всему проекту использовать продвинутые техники при возврате ошибок и их дальнейшую обработку.

ekubyshin commented on internal/config/config.go:
```diff
@@ -5,16 +5,35 @@ import (
+	// ...
+)
+
+var (
```
тут хорошо все смерджил в одно объявление! Супер

ekubyshin commented on internal/db/db.go:
```diff
@@ -0,0 +1,78 @@
+var table string = `
+CREATE TABLE IF NOT EXISTS strazh (
+	id          SERIAL PRIMARY KEY,
+		uid 		INT ,
```
В постгресе есть тип данны UUID, который как раз и предназначен для подобных целей. ТАкже есть дефолтовое значение, GEN_RANDOM_UUID() https://www.postgresql.org/docs/current/functions-uuid.html
Для удобства можно заменить на него, но не обязательно

ekubyshin commented on internal/db/db.go:
```diff
@@ -0,0 +1,78 @@
+CREATE TABLE IF NOT EXISTS strazh (
+	id          SERIAL PRIMARY KEY,
+		uid 		INT ,
+	origin        TEXT,
```
origin и short не достаточно отражают бизнес цель. Предлагаю переименовать, чтобы прям по названию было сразу понятно, что тут будет origin_url и short_url.

По желанию.

ekubyshin commented on internal/db/db.go:
```diff
@@ -0,0 +1,78 @@
+	_ "github.com/jackc/pgx/v5/stdlib"
+)
+
+var table string = `
```
string можно опускать, но можно и оставить.

ekubyshin commented on internal/db/db.go:
```diff
@@ -0,0 +1,78 @@
+)`
+
+/*
+var table string = `
```
Есть ли какая-то функциональная нагрузка у данного sql запроса?

ekubyshin commented on internal/db/db.go:
```diff
@@ -0,0 +1,78 @@
+
+func Ping(ctx context.Context) error {
+
+	err := db.PingContext(ctx)
```
Тут можно сократить до return db.PingContext(ctx)

ekubyshin replied:
Предлагаю сделать красиво

ekubyshin commented on internal/db/db.go:
```diff
@@ -0,0 +1,78 @@
+
+func Insert(ctx context.Context, short, origin string, UUID int) error {
+
+	_, err := db.ExecContext(ctx, "insert into strazh (uid,origin, short) values ($1, $2, $3)", UUID, origin, short)
```
Здесь все хорошо. Если есть желание, то можно все запросы вынести в отдельный файл, а тут просто использовать эти переменные вместо строки.

ekubyshin commented on internal/db/db.go:
```diff
@@ -0,0 +1,78 @@
+
+func Del(ctx context.Context, short string) error {
+
+	_, err := db.ExecContext(ctx, "UPDATE strazh SET deleted = true WHERE short = $1;", short)
```
верно

ekubyshin commented on internal/middleware/withAuth.go:
```diff
@@ -0,0 +1,36 @@
+			fmt.Println("auth err: ", err)
+			jwt, _ := auth.BuildJWTString()
+			userid, _ = auth.GetUserID(jwt)
+			ctx = context.WithValue(ctx, config.JWTkey, string(jwt))
```
Этот момент не совсем понятен. Тут jwt - string,  true - bool, компилятор об этом знает. Зачем дополнительно их приводить в string и bool?

ekubyshin commented on internal/middleware/withAuth.go:
```diff
@@ -0,0 +1,36 @@
+			jwt, _ := auth.BuildJWTString()
+			userid, _ = auth.GetUserID(jwt)
+			ctx = context.WithValue(ctx, config.JWTkey, string(jwt))
+			ctx = context.WithValue(ctx, config.NEWkey, bool(true))
```
Тут на мой взгляд можно создать структуру из двух полей и складывать за один раз с одним ключом.
по желанию

ekubyshin commented on internal/middleware/withGzip.go:
```diff
@@ -0,0 +1,46 @@
+		// проверяем, что клиент умеет получать от сервера сжатые данные в формате gzip
+		acceptEncoding := r.Header.Get("Accept-Encoding")
+		supportsGzip := strings.Contains(acceptEncoding, "gzip")
+		if supportsGzip {
```
здесь можно воспользоваться паттерном guard, те если условие не удовлетворено, то нет смысла делать что то дальше и тогда отпанет необходимость в ow.
если не поддерживается сжатие, то можно сразу передавать выполнение дальше, а если поддерживатся, то делать всю оставшуюся логику.
Как считаешь?

ekubyshin commented on internal/middleware/withlogging.go:
```diff
@@ -0,0 +1,79 @@
+
+type (
+	// берём структуру для хранения сведений об ответе
+	responseData struct {
```
отлично, стиль объявления соблюдается практически везде!

ekubyshin commented on internal/middleware/withlogging.go:
```diff
@@ -0,0 +1,79 @@
+		t1 := time.Now()
+
+		responseData := &responseData{
+			status: 0,
```
по умолчанию будет заполняться нулями при объявлении responseData := &responseData{}

ekubyshin commented on internal/middleware/withlogging.go:
```diff
@@ -0,0 +1,79 @@
+		}
+		h(&lw, r)
+
+		logger.Log.Info("got incoming HTTP request",
```
Тут отлично используешь логгер, а выше использовал fmt.
Предлагаю придерживаться одного способа логирования

ekubyshin commented on internal/server/handlers.go:
```diff
@@ -0,0 +1,375 @@
+*/
+func CreateShortURL(w http.ResponseWriter, req *http.Request) {
+
+	if req.Method == http.MethodPost {
```
Здесь будет удобнее использовать guard паттерн
if req.Method != http.MethodPost {
w.WriteHeader(http. StatusMethodNotAllowed)
return
}

ekubyshin commented on internal/server/handlers.go:
```diff
@@ -0,0 +1,375 @@
+
+		body, err := io.ReadAll(req.Body)
+		if err != nil {
+			return
```
Клиент не поймет, что произошла ошибка. Предлагаю детализировать и возвращать соответствующий статус, например 500

ekubyshin commented on internal/server/handlers.go:
```diff
@@ -0,0 +1,375 @@
+
+		fl := false
+		var conflict *services.ErrConflict409
+		if err == nil {
```
Логика сложна для понимания. Можно воспользоваться guard паттерном и избавиться от лишних конструкций, который усложняют чтение кода.

ekubyshin commented on internal/server/handlers.go:
```diff
@@ -0,0 +1,375 @@
+			fl = true
+			w.WriteHeader(http.StatusCreated)
+		} else {
+			logger.Log.Error(err.Error())
```
здесь отлично - использовал логгер

ekubyshin commented on internal/server/handlers.go:
```diff
@@ -0,0 +1,375 @@
+
+	var deleted *services.ErrDeleted
+
+	if req.Method == http.MethodGet {
```
не забываем про guard pattern

ekubyshin commented on internal/server/handlers.go:
```diff
@@ -0,0 +1,375 @@
+			w.Header().Add("Location", res)
+			w.WriteHeader(http.StatusTemporaryRedirect)
+
+			w.Write([]byte(res))
```
можно сразу сделать return и избавиться от else

ekubyshin commented on internal/server/handlers.go:
```diff
@@ -0,0 +1,375 @@
+		} else if errors.As(err, &deleted) {
+			fmt.Println("запрос удаленного шорта")
+			//w.Header().Add("Location", res)
+			w.WriteHeader(http.StatusGone)
```
Хороший статус код!

ekubyshin commented on internal/server/handlers.go:
```diff
@@ -0,0 +1,375 @@
+
+func CreateShortURLfromJSON(w http.ResponseWriter, req *http.Request) {
+
+	var msg URLmessage
```
не забываем про объединение объявлений

ekubyshin commented on internal/server/handlers.go:
```diff
@@ -0,0 +1,375 @@
+
+	var buf bytes.Buffer
+
+	_, err := buf.ReadFrom(req.Body)
```
бывает так, что боди может быть пустым, можно проверить, что он не пустой и вернуть 400

ekubyshin commented on internal/server/handlers.go:
```diff
@@ -0,0 +1,375 @@
+		return
+	}
+
+	if err = json.Unmarshal(buf.Bytes(), &msg); err != nil {
```
гард паттерн в действии! супер!

ekubyshin commented on internal/server/handlers.go:
```diff
@@ -0,0 +1,375 @@
+	res := config.BaseURL + "/"
+	var resp []byte
+
+	short, err := services.Short(url, -1)
```
маджик намберс ;) чтобы это значило
Можно воспользоваться ENUM или type alias, чтобы передавать не магические числа, а именованые константы на вход

ekubyshin commented on internal/server/handlers.go:
```diff
@@ -0,0 +1,375 @@
+	fl := false
+	var header int
+	var conflict *services.ErrConflict409
+	if err == nil {
```
Тут получилось замысловато ;) Мне пришлось задуматься.
Представь, что твой код будет смотреть другой разработчик, как думаешь, можно ли упроситить этот код отсюда и ниже, чтобы он не был таким громоздким?
Например, нет ничего страшного, если использовать гард и возвращать нужный ответ сразу, а не выставлять флаги и переменные, чтобы вернуть один раз ответ.

ekubyshin commented on internal/server/handlers.go:
```diff
@@ -0,0 +1,375 @@
+	}
+
+	vv, ok := req.Context().Value(config.NEWkey).(bool)
+	fmt.Println("new: ", vv)
```
не забываем, что есть логгер

ekubyshin commented on internal/server/handlers.go:
```diff
@@ -0,0 +1,375 @@
+
+	vv, ok := req.Context().Value(config.NEWkey).(bool)
+	fmt.Println("new: ", vv)
+	if ok && vv {
```
гард паттерн! красота

ekubyshin commented on internal/server/handlers.go:
```diff
@@ -0,0 +1,375 @@
+		w.Write([]byte{})
+		return
+	} else {
+		fmt.Println(ok)
```
как думаешь, можно ли тут просто убрать else и распечатать сразу в лог?

ekubyshin commented on internal/server/handlers.go:
```diff
@@ -0,0 +1,375 @@
+	}
+
+	res := []URLs{}
+	for _, sh := range *services.GetAll() {
```
что здесь произойдет, если GetAll вернет nil?

ekubyshin commented on internal/server/handlers.go:
```diff
@@ -0,0 +1,375 @@
+		fmt.Println(ok)
+	}
+
+	res := []URLs{}
```
для улучшения быстродействия можно создать сразу слайс определенного размера len(services.GetAll())

ekubyshin commented on internal/server/handlers.go:
```diff
@@ -0,0 +1,375 @@
+		//	w.Header().Set("Content-Type", "application/json")
+		w.WriteHeader(http.StatusNoContent)
+
+		w.Write([]byte{})
```
кажется, здесь чего то не хватает, чтобы гард отработал как надо

ekubyshin commented on internal/server/handlers.go:
```diff
@@ -0,0 +1,375 @@
+	if err != nil {
+		logger.Log.Error(err.Error())
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
```
а тут все верно!

ekubyshin commented on internal/server/handlers.go:
```diff
@@ -0,0 +1,375 @@
+		return
+	}
+
+	if req.Method == http.MethodGet {
```
подскажи, этот хэндлер вызывается как в Post, так и в Get?

ekubyshin replied:
Давай разобьем чтобы get возвращал, а пост писал

ekubyshin commented on internal/server/handlers.go:
```diff
@@ -0,0 +1,375 @@
+
+func Ping(w http.ResponseWriter, req *http.Request) {
+
+	if req.Method == http.MethodGet {
```
можно использовать гард паттерн

ekubyshin commented on internal/server/handlers.go:
```diff
@@ -0,0 +1,375 @@
+			w.WriteHeader(http.StatusInternalServerError)
+		}
+
+		w.Write([]byte{})
```
а это для чего?

ekubyshin commented on internal/server/handlers.go:
```diff
@@ -0,0 +1,375 @@
+func Ping(w http.ResponseWriter, req *http.Request) {
+
+	if req.Method == http.MethodGet {
+		ctx, cancel := context.WithTimeout(req.Context(), 1*time.Second)
```
предлагаю подобные магические переменные выносить в константы. В данном случае это похоже, что при запуске приложения можно объявить таймаут обработки запроса и при необходимости использовать его во всех хэндлерах.

ekubyshin replied:
Но то, что вспомнил про отмену операций через контекст - супер!

ekubyshin commented on internal/server/handlers.go:
```diff
@@ -0,0 +1,375 @@
+
+func Batcher(w http.ResponseWriter, req *http.Request) {
+
+	if req.Method == http.MethodPost {
```
не забываем, что гард паттерн значильно улучшает читабельность кода

ekubyshin commented on internal/server/handlers.go:
```diff
@@ -0,0 +1,375 @@
+		var buf bytes.Buffer
+
+		_, err := buf.ReadFrom(req.Body)
+		if err != nil {
```
вот он красавчик

ekubyshin commented on internal/server/handlers.go:
```diff
@@ -0,0 +1,375 @@
+			return
+		}
+
+		if err := json.Unmarshal(buf.Bytes(), &batches); err != nil {
```
и тут тоже красиво с гардом

ekubyshin commented on internal/server/handlers.go:
```diff
@@ -0,0 +1,375 @@
+			return
+		}
+		ll := len(batches)
+		batchback := make([]BatchBack, ll)
```
можно не добавлять лишнюю переменную, а сразу сделать make([]BatchBack, len(batches) код от этого в читабельности не проиграет

ekubyshin replied:
супер, что вспомнил о производительности слайсов и заранее создал слайс нужного размера

ekubyshin commented on internal/server/handlers.go:
```diff
@@ -0,0 +1,375 @@
+		var conflict *services.ErrConflict409
+		for i, b := range batches {
+
+			ress, err := services.Short(b.OriginalURL, -1)
```
магия в действии ))

ekubyshin commented on internal/server/handlers.go:
```diff
@@ -0,0 +1,375 @@
+		for i, b := range batches {
+
+			ress, err := services.Short(b.OriginalURL, -1)
+			if err == nil || errors.As(err, &conflict) {
```
давай вспомним, чем errors.As, отличается от errors.Is?

ekubyshin commented on internal/server/handlers.go:
```diff
@@ -0,0 +1,375 @@
+	_, err := buf.ReadFrom(req.Body)
+
+	if err != nil {
+		//	fmt.Println("err 1: ", err)
```
Предлагаю договоримся, чтобы на ревью передавать код без закомментированных кусков

ekubyshin replied:
Наличие закомментированного кода усложняет проведение код-ревью

ekubyshin commented on internal/server/handlers.go:
```diff
@@ -0,0 +1,375 @@
+
+		w.WriteHeader(http.StatusUnauthorized)
+
+		w.Write([]byte{})
```
кажется, что это лишнее, как считаешь?

ekubyshin commented on internal/server/handlers.go:
```diff
@@ -0,0 +1,375 @@
+	w.Header().Add("Content-Type", "application/json")
+	w.WriteHeader(http.StatusAccepted)
+
+	w.Write([]byte{})
```
можно не писать пустой боди

ekubyshin commented on internal/server/handlers.go:
```diff
@@ -0,0 +1,375 @@
+		return
+	}
+
+	go services.Delete(todel, user)
```
Что делает данный вызов? почему он в горутине?
А как понять, что тут все отработало и не было ошибки?

ekubyshin replied:
Что здесь произойдет, если на сервис будет поступать одновременно много запросов, допустим 1000 запросов в секунду, а работа с бд будет достаточно медленной, чтобы неуспеть удалить урл к тому моменту, когда придет новый запрос с такими же данными?

ekubyshin commented on internal/server/router.go:
```diff
@@ -0,0 +1,23 @@
+	r := chi.NewRouter()
+
+	r.Post(`/`, middleware.AuthMiddleware(middleware.RequestLogger(middleware.GzipMiddleware(CreateShortURL))))
+	r.Post(`/api/shorten`, middleware.RequestLogger(CreateShortURLfromJSON))
```
в чи есть возможность объединять роуты в группы, на мой взгляд это будет прям хорошо. Как считаешь?

ekubyshin commented on internal/server/router.go:
```diff
@@ -0,0 +1,23 @@
+
+	r := chi.NewRouter()
+
+	r.Post(`/`, middleware.AuthMiddleware(middleware.RequestLogger(middleware.GzipMiddleware(CreateShortURL))))
```
действительно ли необходимо на каждый роут объявлять миддлваре? можно ли просто использовать r.Use?

ekubyshin commented on internal/services/conveer.go:
```diff
@@ -0,0 +1,47 @@
+	}
+	var tags []string
+
+	_, err := db.Get().Exec("UPDATE strazh SET deleted = ? WHERE id = ?", true,
```
Тут передается массив id, а  в where используется =.
Если я правильно поняял, то тут необходимо выставить deleted = true, для всех записей из массива id, тогда тут подойдет оператор where id in https://www.postgresqltutorial.com/postgresql-tutorial/postgresql-in/

ekubyshin replied:
а что будет, если попытаться апдейтить запись, которой нет в таблице? вернется ли ошибка?

ekubyshin commented on internal/services/conveer.go:
```diff
@@ -0,0 +1,47 @@
+
+func Delete(shorts []string, uid int) {
+	channels := make([]chan int, len(shorts))
+	for i, short := range shorts {
```
код довольно усложнен, можно ли здесь просто сразу писать результат в буферизованный массив chan int?
Также вспомни про sync.WaitGroup - нужно ли здесь использовать ее?

ekubyshin replied:
Код воспринимается сложно изза переусложненных конструкций.
Что тут происходит?Это реализация fanIn?
Тот код, который выполняется в горутинах имеет ли действительно смысл выполнять в горутинах?
Мб стоит вынести работу с бд в горутины?

ekubyshin commented on internal/services/service.go:
```diff
@@ -1,13 +1,39 @@
-var mapa map[string]string
+var shortuser map[string]*list.List
+
+type ErrConflict409 struct {
```
Здесь очень хорошо вспомнил урок про обработку ошибок и свои типы ошибок. Здорово!

ekubyshin commented on internal/services/service.go:
```diff
@@ -1,13 +1,39 @@
+	return e.s
+}
 
 func generateRandomString(length int) string {
```
Эх, тесты бы на эту замечательную функцию!

ekubyshin commented on internal/services/service.go:
```diff
@@ -21,39 +47,70 @@ func generateRandomString(length int) string {
 }
 
-func Short(url string) string {
+func Short(oririn string, uuid int) (string, error) {
```
очепятка ;)

ekubyshin replied:
данная функция вызывается в нескольких потоках?

ekubyshin commented on internal/services/service.go:
```diff
@@ -21,39 +47,70 @@ func generateRandomString(length int) string {
-	if ok {
-		fmt.Println(url, " уже есть в мапе: ", v)
-		return v
+			return sh.ShortURL, &ErrConflict409{config.Conflict409}
```
красота в гардах!

ekubyshin commented on internal/services/service.go:
```diff
@@ -21,39 +47,70 @@ func generateRandomString(length int) string {
+func Short(oririn string, uuid int) (string, error) {
+
+	//v, ok := mapa[oririn]
+	for _, sh := range shorten {
```
а что такое shorten и откуда он взялся?

ekubyshin commented on internal/services/service.go:
```diff
@@ -21,39 +47,70 @@ func generateRandomString(length int) string {
 	}
+	fmt.Println("Short ", len(shorten), " ", id)
+	sh := storager.Shorten{ID: id, UUID: uuid, ShortURL: short, OriginalURL: oririn, DeletedFlag: false}
+	shorten = append(shorten, &sh)
```
не понимаю назначение этой глобальной переменной.  Объясни, пожалуйста.

ekubyshin commented on internal/storager/dbstorager.go:
```diff
@@ -0,0 +1,49 @@
+type DBStorager struct {
+}
+
+func (one *DBStorager) Load(shorten *[]*Shorten) error {
```
стоит ли передавать ссылку на слайс? Давай вспомним, что из себя представляет слайс

ekubyshin commented on internal/storager/dbstorager.go:
```diff
@@ -0,0 +1,49 @@
+
+func (one *DBStorager) Load(shorten *[]*Shorten) error {
+
+	rows, err := db.Get().Query("SELECT * FROM strazh")
```
Подскажи, надо ли тут делать defer rows.Close()?

ekubyshin replied:
Здесь пропустил два момента
1. Предлагаю указать вместо * конкретные названия столбцов, которые выбираем, тк это позволит избежать ошибок в будущем, когда добавится новый столбец или удалиться.
2. посмотри, пожалуйста, внимательно документацию https://pkg.go.dev/database/sql#Rows.Close необходимо добавить закрытие defer rows.Close()

ekubyshin commented on internal/storager/dbstorager.go:
```diff
@@ -0,0 +1,49 @@
+
+func (one *DBStorager) Load(shorten *[]*Shorten) error {
+
+	rows, err := db.Get().Query("SELECT * FROM strazh")
```
Выбирать все поля через * является не желательной практикой, тк настоящие сервисы развиваются, добавляются новые колонки, какие-то могут удаляться со временем. Поэтому если выбирать поля через *, то в какой-то момент код может начать падать в продакшене, потому что полей вернется больше, чем было или наоборот меньше. Поэтому * является довольно рискованной выборкой с последующим мапингом.

ekubyshin commented on internal/storager/dbstorager.go:
```diff
@@ -0,0 +1,49 @@
+
+func (one *DBStorager) StorageWrite(short, origin string, UUID int) error {
+
+	return db.Insert(context.Background(), short, origin, UUID)
```
Помнишь в теме где изучали работу с бд, рассматривали вопрос передачи контекста?
Хороше практикой, является передача контекста при работе с бд и другими местами, где потом потребуется отмена операции по таймауту. Здесь уже не получится отменить операцию по таймауту, а если запрос зависнет, то будет неожиданное поведение.

ekubyshin commented on internal/storager/dbstorager.go:
```diff
@@ -0,0 +1,49 @@
+func (one *DBStorager) Init() error {
+	ctx, cancel := context.WithTimeout(context.Background(), 1*time.Second)
+	defer cancel()
+	return db.Ping(ctx)
```
а тутвсе хорошо с контекстом

ekubyshin commented on internal/storager/dbstorager.go:
```diff
@@ -0,0 +1,49 @@
+}
+
+func (one *DBStorager) Delete(listID []int) {
+	fmt.Println("DBStorager delete ", listID)
```
хмм. какое предназначение у этого метода?

ekubyshin commented on internal/storager/filestorager.go:
```diff
@@ -0,0 +1,89 @@
+	if err != nil {
+		return err
+	}
+	defer file.Close()
```
не забыл закрыть файл - супер!

ekubyshin commented on internal/storager/filestorager.go:
```diff
@@ -0,0 +1,89 @@
+		return err
+	}
+
+	err = writer.Flush()
```
Как думаешь, нужно ли потом закрывать writer?

GLZG replied:
Я не нашел в интерфейсе bufio.Writer отдельной команды для закрытия буфера
Считаю, что после завершения работы с файлом нужно закрыть обьект os.File:

	defer file.Close()

Сам же writer как локальная переменная уничтожится сборщиком мусора

ekubyshin commented on internal/storager/interface.go:
```diff
@@ -0,0 +1,50 @@
+	}
+	logger.Log.Error(err.Error())
+	store = &FileStorager{}
+	if err = store.Init(); err == nil {
```
Предлагаю воспользоваться конструктором NewFileStorage, который сделает все что нужно и вернет объект.;
Предпочтительнее использовать функции конструкторы при создании объектов

GLZG replied:
Выполнил

ekubyshin commented on pkg/compress/compress.go:
```diff
@@ -0,0 +1,109 @@
+		return nil, fmt.Errorf("failed write data to compredd temporary buffer: %v ", err)
+	}
+
+	err = w.Close()
```
Отлично!

ekubyshin commented on pkg/compress/compress.go:
```diff
@@ -0,0 +1,109 @@
+
+func Decompress(data []byte) ([]byte, error) {
+	r := flate.NewReader(bytes.NewReader(data))
+	defer r.Close()
```
тоже отлично

ekubyshin commented on pkg/compress/compress_test.go:
```diff
@@ -0,0 +1,33 @@
+	"github.com/stretchr/testify/require"
+)
+
+func TestCompress(t *testing.T) {
```
тесты это хорошо! Особенно когда их много! ;)

ekubyshin commented on internal/config/config.go:
```diff
@@ -5,16 +5,35 @@ import (
 
-var BaseURL string
+const (
+	UIDkey key = iota
```
неожиданно, что вспомнил про iota! Не каждый разработчик вспоминает об этой возможности

Commit:
.

Commit:
increment userid

Commit:
const userid

Commit:
переместил TOKENEXP и SECRETKEY в config

Commit:
Заменил fmt на logger
убрал комменты

Commit:
.

Commit:
Merge remote-tracking branch 'origin/iter15' into iter15

Commit:
.

Commit:
добавил тип timeError

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
убрал dot

Commit:
var ErrBuildJWTString error = errors.New("ошибка формирования JWT")

Commit:
move handlers to own files

Commit:
.

Commit:
Merge remote-tracking branch 'origin/iter15' into iter15

Commit:
add middleware for errors handling

Commit:
.

Commit:
убрал логирование ошибок из хэндлеров
пробросил ошибки из сервсиа через хэндлеры в мидл обработчик ошибок

ekubyshin commented on internal/errors/timeError.go:
```diff
@@ -0,0 +1,25 @@
+)
+
+// TimeError предназначен для ошибок с фиксацией времени возникновения.
+type TimeError struct {
```
верно

ekubyshin commented on internal/logger/logger.go:
```diff
@@ -0,0 +1,29 @@
+	if err != nil {
+		return err
+	}
+	// устанавливаем синглтон
```
Паттерны проходятся дальше по курсу. Синглтон чуть иначе реализуется, но в данный момент оставим так.

ekubyshin commented on internal/middleware/Auth.go:
```diff
@@ -0,0 +1,38 @@
+	return func(w http.ResponseWriter, r *http.Request) (err error) {
+		// по умолчанию устанавливаем оригинальный http.ResponseWriter как тот,
+		// который будем передавать следующей функции
+		ow := w
```
тут так и не понял смысл создания локального ow

GLZG replied:
исправил

ekubyshin commented on internal/middleware/ErrHandler.go:
```diff
@@ -0,0 +1,18 @@
+	return func(w http.ResponseWriter, r *http.Request) {
+		err := f(w, r)
+		if err != nil {
+			logger.Log.Error("err: ", zap.String("", err.Error()))
```
интересное решение с мидлвар

ekubyshin commented on internal/server/Batcher.go:
```diff
@@ -0,0 +1,68 @@
+
+func Batcher(w http.ResponseWriter, req *http.Request) error {
+
+	if req.Method == http.MethodPost {
```
Здесь не доработал немного, эту конструкцию можно упростить через паттерн guard.
Паттерн предполагает, что в самом начале проверяется условие, при котором выполнение не имеет смысла и прерывается выполнение.
Здесь можно убрать этот большой if и поменять условие на отрицание с возвратом из метода

if req.Method != http.MethodPost { 
w.WriteHeader(http.StatusBadRequest) 
return
}

GLZG replied:
Поправил в 8434f92

ekubyshin commented on internal/server/CreateShortURL.go:
```diff
@@ -0,0 +1,67 @@
+*/
+func CreateShortURL(w http.ResponseWriter, req *http.Request) error {
+
+	if req.Method == http.MethodPost {
```
Здесь, также забыл про guard паттерн.
Предлагаю переделать, чтобы закрепить применение этого паттерна на практике

ekubyshin commented on internal/server/Delete.go:
```diff
@@ -0,0 +1,47 @@
+	var todel []string
+	var buf bytes.Buffer
+
+	_, err = buf.ReadFrom(req.Body)
```
Здесь используется buf.ReadFrom, в другом месте используется io.ReadAll.
Давай приведем все к одному виду, предлагаю io.ReadAll
или вот еще один способ https://golangbyexample.com/json-request-body-golang-http/
с использованием декодера. 
При работе над проектом желательно придерживаться одного стиля написания кода.
Предлагаю исправить.

ekubyshin commented on internal/server/GetURL.go:
```diff
@@ -0,0 +1,59 @@
+
+	var deleted *services.ErrDeleted
+
+	if req.Method == http.MethodGet {
```
Вспоминаем про guard pattern

ekubyshin commented on internal/server/GetURLs.go:
```diff
@@ -0,0 +1,70 @@
+		//	w.Header().Set("Content-Type", "application/json")
+		w.WriteHeader(http.StatusNoContent)
+
+		w.Write([]byte{})
```
Имеет ли смысл коду ниже выполняться, если уже записал пустой боди? 
Стоит ли писать пустой респонс, если по умолчанию он и так будет пустым?
Предлагаю исправить

GLZG replied:
поправил в 05c8be7

ekubyshin commented on internal/server/router.go:
```diff
@@ -0,0 +1,23 @@
+
+	r := chi.NewRouter()
+
+	r.Post(`/`, middleware.ErrHandler(middleware.Auth(middleware.Log(middleware.Gzip(CreateShortURL)))))
```
Предлагаю навести красоту и воспользоваться 
https://go-chi.io/#/pages/routing?id=routing-groups
А также можно использовать middleware сразу для всего роута r.Use(.....)

GLZG replied:
Не могу вставить  middleware.ErrHandler аргументом в r.Use

../../internal/server/router.go:15:9: cannot use middleware.ErrHandler (value of type func(f middleware.MyHandlerFunc) http.HandlerFunc) as func(http.Handler) http.Handler value in argument to r.Use

ekubyshin commented on internal/services/conveer.go:
```diff
@@ -0,0 +1,51 @@
+	}
+
+	str := "UPDATE strazh SET deleted = true WHERE id IN (" + strings.Join(listID, ",") + ")"
+	fmt.Println("query: ", str)
```
Предлагаю избавиться по всему проект от fmt.Prinln,  тк уже есть логгер

ekubyshin commented on internal/services/conveer.go:
```diff
@@ -0,0 +1,51 @@
+		ch := make(chan int)
+		go func(short string, uid int, ch chan int) {
+			id := -1
+			for _, sh := range shorten {
```
по коду не понятно откуда взялся shorten и чем он занимается, что в нем хранится. 
предлагаю поработать над методом удаления и упростить его реализацию

GLZG replied:
Переделал метод Delete в f00c2b8

ekubyshin commented:
Еще стоит уделить внимание сообщениям в каждом коммите. Понимаю, что это кажется не важным и код говорит сам за себя, но в работе над реальным проеком, в сообщениях полезно указывать комментарий, который отражает суть изменений в конкретном коммите. Давай поработаем над этим.

Commit:
//fmt.Println

Commit:
.

Commit:
паттерн guard в Batcher

Commit:
return from func GetUrl after writing StatusNoContent header

Commit:
оказалось что нельзя взять и положить видоизмененную миддлу в метод Use

Commit:
try to put error in context

Commit:
bunrouter

Commit:
Merge remote-tracking branch 'origin/iter15' into iter15

Commit:
.

Commit:
.

Commit:
.

Commit:
переделал Delete

Commit:
Merge remote-tracking branch 'origin/iter15' into iter15

Commit:
Revert "Merge remote-tracking branch 'origin/iter15' into iter15"

This reverts commit 24b1a1d148fb728766d3f6f842f3674d404d95b0, reversing
changes made to 117d7dab091eb7a5586109a907c3d2d78d5d19ca.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
привел чтение из req.Body к одному стилю

Commit:
add generateRandomString_test.go

Commit:
в мидле errHandler обьявляю ошибку и кладу в контекст реквеста указатель на нее
в хэндлерах и мидлах кладу ошибки в контекст реквеста AddErrToReqContext
после в мидле errHandler обрабатываю

Commit:
C учтом того, что AddErrToReqContext вызывается через defer, теперь передаю в него вместо ошибки err указатель на нее *err

Commit:
подчистил ненужные return

Commit:
подчистил все ненужные return.

ekubyshin commented on internal/config/config.go:
```diff
@@ -5,16 +5,44 @@ import (
+	SECRETKEY string
+)
+
+const Conflict409 = "попытка сократить уже имеющийся в базе URL"
```
отлично

ekubyshin commented on internal/server/Batcher.go:
```diff
@@ -0,0 +1,71 @@
+	defer packerr.AddErrToReqContext(req, &err)
+
+	logger.Log.Info("Batcher")
+	if req.Method != http.MethodPost {
```
вот так!

ekubyshin commented on internal/server/GetURL.go:
```diff
@@ -0,0 +1,59 @@
+		w.Header().Add("Location", res)
+		w.WriteHeader(http.StatusTemporaryRedirect)
+
+		w.Write([]byte(res))
```
здесь можно поставить просто return и избавиться от else

ekubyshin commented on internal/server/GetURL.go:
```diff
@@ -0,0 +1,59 @@
+
+		w.Write([]byte(res))
+
+	} else if errors.As(err, &deleted) {
```
здесь else будет  не нужен

ekubyshin commented on internal/server/GetURL.go:
```diff
@@ -0,0 +1,59 @@
+		//w.Header().Add("Location", res)
+		w.WriteHeader(http.StatusGone)
+
+		w.Write([]byte(res))
```
return

ekubyshin commented on internal/server/GetURL.go:
```diff
@@ -0,0 +1,59 @@
+		w.WriteHeader(http.StatusGone)
+
+		w.Write([]byte(res))
+	} else {
```
else можно убрать, тк если в if стоит return, то дальше код не будет выболняться. А все что ниже уже по умолчанию станет else

ekubyshin commented on internal/server/Ping.go:
```diff
@@ -0,0 +1,29 @@
+	defer cancel()
+	if err := db.Ping(ctx); err == nil {
+		w.WriteHeader(http.StatusOK)
+
```
return

ekubyshin commented on internal/server/Ping.go:
```diff
@@ -0,0 +1,29 @@
+	if err := db.Ping(ctx); err == nil {
+		w.WriteHeader(http.StatusOK)
+
+	} else {
```
можно убрать

ekubyshin commented on internal/server/router.go:
```diff
@@ -0,0 +1,26 @@
+func InitRouter() {
+
+	r := chi.NewRouter()
+	r.Group(func(r chi.Router) {
```
Тут не совсем корректно используется r.Group, 
Давай посмотрим документацию
`
r.Route("/articles", func(r chi.Router) {
    r.With(paginate).Get("/", listArticles)                           // GET /articles
    r.With(paginate).Get("/{month}-{day}-{year}", listArticlesByDate) // GET /articles/01-16-2017

    r.Post("/", createArticle)                                        // POST /articles
    r.Get("/search", searchArticles)                                  // GET /articles/search

    // Regexp url parameters:
    r.Get("/{articleSlug:[a-z-]+}", getArticleBySlug)                // GET /articles/home-is-toronto

    // Subrouters:
    r.Route("/{articleID}", func(r chi.Router) {
      r.Use(ArticleCtx)
      r.Get("/", getArticle)                                          // GET /articles/123
      r.Put("/", updateArticle)                                       // PUT /articles/123
      r.Delete("/", deleteArticle)                                    // DELETE /articles/123
    })
  })

`

В твоем пример будет так

r.Group('/api',func(r chi.Router) {
r.Post('/shorten/batch',....)
r.Post('/shorten'....)
r.Get('/user/urls',....)
r.Delete('user/urls',...)
тут можно даже объединять еще

тк мидлвар лог подключается везде, то можно его подключать один раз на все
r.Use(middleware.ErrHandler)
r.Use(middleware.Log)
r.Use(middleware.Auth) - тут можно уже подключить внутри группы, хотя на мой взгляд создание урлов тоже должно быть под авторизацией.
})

ekubyshin commented on internal/auth/auth.go:
```diff
@@ -0,0 +1,67 @@
+	return tokenString, nil
+}
+
+func GetUserID(tokenString string) (int, error) {
```
Посмотри, пожалуйста, внимательнее задание Инкремент 14, в разделе Безопасность. Там было указано, чтобы устанавливать авторизационную куку, после прохождения успешной авторизации.
`Выдавать пользователю симметрично подписанную куку, содержащую уникальный идентификатор пользователя, если такой куки не существует или она не проходит проверку подлинности.`
Не увидел этого кода

Commit:
попытка в fanID

Commit:
поправил

Commit:
.

Commit:
поправил

Commit:
Revert "поправил"

This reverts commit ac8d1bff827ad8e19251f25cbb5e2437955151ce.

Commit:
.

Commit:
group routes

Commit:
Revert "group routes"

This reverts commit f0c7970bb8cdbbb206bcbb2bf9a26a73dcb0de53.

Commit:
.

Commit:
rotate Auth and Log

Commit:
.

Commit:
group

Commit:
group

Commit:
try to add coockie

Commit:
coockie

Commit:
.

Commit:
.

Commit:
.

ekubyshin reviewed (approved):
Как и договаривались - принимаю работу, с учетом, что ошибки исправишь при работе над выпускным проектом.

Поздравляю со сдачей спринта 4
